### PR TITLE
Support gRPC in the Upstreams of the virtual server resources

### PR DESCRIPTION
**Pull request by ciarams87**
### Proposed changes
Support gRPC in the Upstreams of the virtual server resources. 

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/kubernetes-ingress/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto master
- [x] I will ensure my PR is targeting the master branch and pulling from my branch from my own fork

### CODE REVIEW

**Review by pleshakov — COMMENTED** [review]
@ciarams87  please see new feedback

File: internal/configs/version2/nginx-plus.virtualserver.tmpl [hunk 1]
```diff
@@ -363,47 +363,117 @@ server {
         proxy_buffers {{ $l.ProxyBuffers }};
             {{ end }}
             {{ if $l.ProxyBufferSize }}
-        proxy_buffer_size {{ $l.ProxyBufferSize }};
+        {{ $proxyOrGRPC }}_buffer_size {{ $l.ProxyBufferSize }};
             {{ end }}
-        proxy_http_version 1.1;
+        
+        {{ if $l.GRPCPass }}
```
[the remark this answers is not in the document]
Comment (ciarams87) [reply internal/configs/version2/nginx-plus.virtualserver.tmpl:369]: Good catch!

File: tests/suite/test_virtual_server_mixed_grpc.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                               virtual_server_setup.public_endpoint.port_ssl)
+        target = f'{virtual_server_setup.public_endpoint.public_ip}:{virtual_server_setup.public_endpoint.port_ssl}'
+        credentials = grpc.ssl_channel_credentials(root_certificates=cert.encode())
+        options = (('grpc.ssl_target_name_override', virtual_server_setup.vs_host),)
+
+        with grpc.secure_channel(target, credentials, options) as channel:
+            stub = GreeterStub(channel)
+            response = ""
+            try:
+                response = stub.SayHello(HelloRequest(name=virtual_server_setup.public_endpoint.public_ip))
```
Comment (vepatel) [inline tests/suite/test_virtual_server_mixed_grpc.py:122]: is it possible to assert something based on the data in response?

File: tests/suite/test_virtual_server_grpc.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        assert "grpc_set_header X-Forwarded-Proto $scheme;" in config
+
+        assert 'grpc_set_header Host "$host";' in config
+
+        assert "grpc_next_upstream error timeout;" in config
+        assert "grpc_next_upstream_timeout 0s;" in config
+        assert "grpc_next_upstream_tries 0;" in config
+
+        print("\nStep 2: Check connection to app")
+        self.connect_grpc_backend(virtual_server_setup)
```
Comment (vepatel) [inline tests/suite/test_virtual_server_grpc.py:144]: is it possible to assert something based on the data in response?

File: tests/suite/test_v_s_route_grpc.py [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        self.deploy_tls_secrets(kube_apis, v_s_route_setup)
+        print("\nStep 1: assert config")
+        ic_pod_name = get_first_pod_name(kube_apis.v1, ingress_controller_prerequisites.namespace)
+        config = get_vs_nginx_template_conf(kube_apis.v1,
+                                            v_s_route_setup.namespace,
+                                            v_s_route_setup.vs_name,
+                                            ic_pod_name,
+                                            ingress_controller_prerequisites.namespace)
+
+        assert "proxy_connect_timeout 60s;" not in config
```
Comment (vepatel) [inline tests/suite/test_v_s_route_grpc.py:101]: these assertions can be separated out in custom_assertions.py and can be reused in all 3 tests by a single function call.

File: tests/suite/test_virtual_server_grpc.py (hunk 3 above)
Comment (ciarams87) [reply tests/suite/test_virtual_server_grpc.py]: Done

File: tests/suite/test_v_s_route_grpc.py (hunk 4 above)
Comment (ciarams87) [reply tests/suite/test_v_s_route_grpc.py]: Done

File: tests/suite/test_virtual_server_mixed_grpc.py (hunk 2 above)
Comment (ciarams87) [reply tests/suite/test_virtual_server_mixed_grpc.py:99]: Done

File: internal/configs/virtualserver.go [hunk 5]
```diff
@@ -1329,7 +1338,11 @@ func upstreamHasKeepalive(upstream conf_v1.Upstream, cfgParams *ConfigParams) bo
 	return cfgParams.Keepalive != 0
 }
 
-func generateRewrites(path string, proxy *conf_v1.ActionProxy, internal bool, originalPath string) []string {
+func generateRewrites(path string, proxy *conf_v1.ActionProxy, internal bool, originalPath string, grpcEnabled bool) []string {
+	if grpcEnabled && internal {
```
Comment (pleshakov) [inline internal/configs/virtualserver.go:1342]: this is still pending -- https://github.com/nginxinc/kubernetes-ingress/pull/2110#discussion_r741533516

basically, when `grpcEnabled && internal` is `true`, we can still have `proxy != nil && proxy.RewritePath != ""` (meaning there is rewrite configured using the proxy action). In this case we should not return, but rather add `^ $request_uri` (without the break) as a first rewrite in the result.

**Review by pleshakov — COMMENTED** [review]
Hi @ciarams87  there is one pending comment, but otherwise looks good to me! thanks for addressing the feedback 👍

File: internal/configs/virtualserver.go (hunk 5 above)
Comment (ciarams87) [reply internal/configs/virtualserver.go:1342]: Oh I must have missed this one! Thanks 😄 

I've updated that function and added another test for the case where `grpcEnabled && internal` is `true`,and we still have `proxy != nil && proxy.RewritePath != ""`

**Review by pleshakov — APPROVED** [review]

**Review by soneillf5 — APPROVED** [review]
